src/rate.rs

```diff
@@ -703,22 +731,33 @@ fn chroma_offset(
 impl QuantizerParameters {
   fn new_from_log_q(
     log_base_q: i64, log_target_q: i64, bit_depth: usize,
-    chroma_sampling: ChromaSampling,
+    chroma_sampling: ChromaSampling, is_intra: bool,
   ) -> QuantizerParameters {
     let scale = q57(QSCALE + bit_depth as i32 - 8);
-    let quantizer = bexp64(log_target_q + scale);
-    let (offset_u, offset_v) = chroma_offset(log_target_q, chroma_sampling);
+
+    let mut log_q_y = log_target_q;
+    if !is_intra && bit_depth == 8 {
+      log_q_y = log_target_q
+        + (log_target_q >> 32) * Q_MODEL_MUL[chroma_sampling as usize]
+        + Q_MODEL_ADD[chroma_sampling as usize];
+    }
+
+    let quantizer = bexp64(log_q_y + scale);
+    let (offset_u, offset_v) = chroma_offset(log_q_y, chroma_sampling);
     let mono = chroma_sampling == ChromaSampling::Cs400;
-    let log_target_q_u = log_target_q + offset_u;
-    let log_target_q_v = log_target_q + offset_v;
-    let quantizer_u = bexp64(log_target_q_u + scale);
-    let quantizer_v = bexp64(log_target_q_v + scale);
+    let log_q_u = log_q_y + offset_u;
+    let log_q_v = log_q_y + offset_v;
+    let quantizer_u = bexp64(log_q_u + scale);
+    let quantizer_v = bexp64(log_q_v + scale);
     let lambda = (::std::f64::consts::LN_2 / 6.0)
       * ((log_target_q as f64) * Q57_SQUARE_EXP_SCALE).exp();
-    let lambda_u = (::std::f64::consts::LN_2 / 6.0)
-      * ((log_target_q_u as f64) * Q57_SQUARE_EXP_SCALE).exp();
-    let lambda_v = (::std::f64::consts::LN_2 / 6.0)
-      * ((log_target_q_v as f64) * Q57_SQUARE_EXP_SCALE).exp();
+
+    let log_q = [log_q_y, log_q_u, log_q_v];
+    let mut dist_scale = [0.; 3];
+    for plane in 0..3 {
+      dist_scale[plane] =
+        bexp64((log_target_q - log_q[plane]) * 2 + q57(16)) as f64 / 65536.;
+    }
```

Comment:
For the record, we could apply the map-and-collect pattern like so:
```suggestion
    let dist_scale = [log_q_y, log_q_u, log_q_v]
      .iter()
      .map(|q| bexp64((log_target_q - q) * 2 + q57(16)) as f64 / 65536.)
      .collect::<ArrayVec<_>>()
      .into_inner()
      .unwrap();
```

Reply:
How about:
```rust
    let scale = |q| bexp64((log_target_q - q) * 2 + q57(16)) as f64 / 65536.;
    let dist_scale = [scale(log_q_y), scale(log_q_u), scale(log_q_v)];
```